File: 2-DataStructures/2-Linked-Lists/1-Single-Linked-Lists/single-linked-lists.py
# Singly Linked Lists definition: A singly linked list is a collection of nodes that collectively form a linear sequence.
# Each node stores a reference to an object that is an element of the sequence, as well as a reference to the next node of the list.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __repr__(self):
        return str(self.data)


class SingleLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def insert(self, data):
        self.size += 1
        new_node = Node(data)

        if not self.head:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head = new_node

    # O(1) time complexity

    def inser_start(self, data):
        # self.num_nodes += 1
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            # self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        self.length += 1

    # O(n) time complexity
    def insert_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current_node = self.head

            # this is why it has O(n) time complexity
            while current_node.next is not None:
                current_node = current_node.next

            current_node.next = new_node
        self.length += 1

    # O(n) time complexity
    def insert_at(self, data, position):
        new_node = Node(data)
        if position == 0:
            self.inser_start(data)
        elif position == self.length:
            self.insert_end(data)
        else:
            current_node = self.head
            current_position = 0
            while current_position < position - 1:
                current_node = current_node.next
                current_position += 1

            new_node.next = current_node.next
            current_node.next = new_node
            self.length += 1

    def remove(self, data):
        if self.head is None:
            return None

        actual_node = self.head
        previous_node = None

        while actual_node is not None and actual_node.data != data:
            previous_node = actual_node
            actual_node = actual_node.next

        if actual_node is None:
            return

        if previous_node is None:
            self.head = actual_node.next
        else:
            previous_node.next = actual_node.next  # GC will remove the node

    # O(n) time complexity
    def traverse(self):
        current_node = self.head
        while current_node is not None:
            print(current_node.data)
            current_node = current_node.next


# Test
my_list = SingleLinkedList()
my_list.inser_start(1)
my_list.inser_start("John")
my_list.inser_start(True)
my_list.insert_end(10.12)
my_list.insert_end("Hello")
my_list.insert_end(100)
my_list.insert_at("Middle", 3)
my_list.remove(100)
my_list.remove("John")
my_list.remove(True)
my_list.traverse()

File: 2-DataStructures/2-Linked-Lists/1-Single-Linked-Lists/single-linked-lists.js
// Linked Lists is a data structure that contains a head, tail and length property.
// Linked Lists consist of nodes, and each node has a value and a pointer to another node or null.
// Linked Lists do not have indexes.
// Linked Lists are an excellent alternative to arrays when insertion and deletion at the beginning are frequently required.
// Arrays contain a built in index whereas Linked Lists do not.
// The idea of a list data structure that consists of nodes is the foundation for other data structures like Stacks and Queues.

const basket = ['apples', 'grapes', 'pears'];
// apples
// 8947 --> grapes
//           8742 --> pears
//                     372 --> null

// playing with pointer references:
let obj1 = { a: true };
let obj2 = obj1;
obj1.a = 'changed';
console.log('obj1', obj1);
console.log('obj2', obj2);
delete obj1;
// console.log('obj1', obj1);
console.log('obj2', obj2); // there is still a pointer to this point in memory

// building a linked list from scratch
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor(value) {
    this.head = new Node(value); // the head and tail are the same node
    this.tail = this.head;
    this.length = 1;
  }

  append(value) { // O(1)
    const newNode = new Node(value);
    this.tail.next = newNode;
    this.tail = newNode;
    this.length++;
    return this;
  }

  prepend(value) { // O(1)
    const newNode = new Node(value);
    newNode.next = this.head;
    this.head = newNode;
    this.length++;
    return this;
  }

  insert(index, value) { // O(n)
    if (index >= this.length) {
      return this.append(value);
    }
    if (index === 0) {
      return this.prepend(value);
    }
    const newNode = new Node(value);
    const leader = this.traverseToIndex(index - 1);
    const holdingPointer = leader.next;
    leader.next = newNode;
    newNode.next = holdingPointer;
    this.length++;
    return this;
  }

  traverseToIndex(index) { // O(n)
    let counter = 0;
    let currentNode = this.head;
    while (counter !== index) {
      currentNode = currentNode.next;
      counter++;
    }
    return currentNode;
  }

  print() {
    const array = [];
    let currentNode = this.head;
    while (currentNode !== null) {
      array.push(currentNode.value)
      currentNode = currentNode.next;
    }
    console.log(array);
  }

  remove(index) { // O(n)
    if (index === 0) {
      this.head = this.head.next;
      this.length--;
      return this;
    }
    if (index >= this.length) {
      index = this.length - 1;
    }
    const leader = this.traverseToIndex(index - 1);
    const unwantedNode = leader.next;
    leader.next = unwantedNode.next;
    this.length--;
    return this;
  }

  reverse() { // O(n)
    if (!this.head.next) {
      return this.head;
    }
    let first = this.head;
    this.tail = this.head;
    let second = first.next;
    while(second) {
      const temp = second.next; // save the next node
      second.next = first; // reverse the pointer
      first = second; // move the pointers forward
      second = temp;
    }
    this.head.next = null;
    this.head = first;
    return this;
  }
}

const myLinkedList = new LinkedList(10);
myLinkedList.append(5);
myLinkedList.append(16);
myLinkedList.prepend(1);
myLinkedList.insert(2, 99);
myLinkedList.remove(2);
myLinkedList.print();
myLinkedList.reverse();
myLinkedList.print();

// 10 --> 5 --> 16
let linkedList = {
  head: {
    value: 10,
    next: {
      value: 5,
      next: {
        value: 16,
        next: null,
      }
    }
  }
}